File: results/0.25/Network-Time-Protocol-Version-4/network-time-protocol-version-4-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    uint8_t li_vn_mode;
    uint8_t stratum;
    uint8_t poll;
    uint8_t precision;
    uint32_t root_delay;
    uint32_t root_dispersion;
    uint32_t reference_id;
    uint64_t reference_timestamp;
    uint64_t origin_timestamp;
    uint64_t receive_timestamp;
    uint64_t transmit_timestamp;
} ntp_packet;

typedef struct {
    uint16_t field_type;
    uint16_t length;
    uint8_t *value;
} extension_field;

typedef struct {
    uint32_t key_id;
    uint8_t *digest;
} mac_field;

static void print_ntp_packet(const HParseResult *result) {
    if (!result || !result->ast) return;

    const HParsedToken *token = result->ast;
    if (token->token_type != TT_SEQUENCE) return;

    printf("LI_VN_MODE: 0x%02x\n", (uint8_t)token->seq->elements[0]->uint);
    printf("Stratum: %u\n", (uint8_t)token->seq->elements[1]->uint);
    printf("Poll: %d\n", (int8_t)token->seq->elements[2]->uint);
    printf("Precision: %d\n", (int8_t)token->seq->elements[3]->uint);
    printf("Root Delay: 0x%08x\n", (uint32_t)token->seq->elements[4]->uint);
    printf("Root Dispersion: 0x%08x\n", (uint32_t)token->seq->elements[5]->uint);
    printf("Reference ID: 0x%08x\n", (uint32_t)token->seq->elements[6]->uint);
    printf("Reference Timestamp: 0x%016llx\n", (unsigned long long)token->seq->elements[7]->uint);
    printf("Origin Timestamp: 0x%016llx\n", (unsigned long long)token->seq->elements[8]->uint);
    printf("Receive Timestamp: 0x%016llx\n", (unsigned long long)token->seq->elements[9]->uint);
    printf("Transmit Timestamp: 0x%016llx\n", (unsigned long long)token->seq->elements[10]->uint);

    if (token->seq->elements[11] && token->seq->elements[11]->token_type != TT_NONE) {
        printf("Optional Fields Present\n");
    }
}

static HParser *create_ntp_parser(void) {
    // Basic fields
    HParser *li_vn_mode = h_uint8();
    HParser *stratum = h_uint8();
    HParser *poll = h_uint8();
    HParser *precision = h_uint8();
    HParser *root_delay = h_uint32();
    HParser *root_dispersion = h_uint32();
    HParser *reference_id = h_uint32();
    HParser *reference_timestamp = h_uint64();
    HParser *origin_timestamp = h_uint64();
    HParser *receive_timestamp = h_uint64();
    HParser *transmit_timestamp = h_uint64();

    // Extension field parser
    HParser *extension_field = h_sequence(
        h_uint16(), // field type
        h_uint16(), // length
        h_length_value(h_uint16(), h_uint8()), // value with padding
        NULL
    );

    // MAC field parser
    HParser *mac_field = h_sequence(
        h_uint32(), // key id
        h_length_value(h_uint16(), h_uint8()), // digest
        NULL
    );

    // Optional fields
    HParser *optional_fields = h_many(h_choice(extension_field, mac_field, NULL));

    // Complete NTP packet
    return h_sequence(
        li_vn_mode,
        stratum,
        poll,
        precision,
        root_delay,
        root_dispersion,
        reference_id,
        reference_timestamp,
        origin_timestamp,
        receive_timestamp,
        transmit_timestamp,
        optional_fields,
        NULL
    );
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <ntp_packet_file>\n", argv[0]);
        return 1;
    }

    FILE *f = fopen(argv[1], "rb");
    if (!f) {
        perror("Failed to open file");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    size_t size = ftell(f);
    fseek(f, 0, SEEK_SET);

    uint8_t *buf = malloc(size);
    if (!buf) {
        fclose(f);
        fprintf(stderr, "Memory allocation failed\n");
        return 1;
    }

    if (fread(buf, 1, size, f) != size) {
        free(buf);
        fclose(f);
        fprintf(stderr, "Failed to read file\n");
        return 1;
    }

    HParser *ntp_parser = create_ntp_parser();
    if (!ntp_parser) {
        free(buf);
        fclose(f);
        fprintf(stderr, "Failed to create parser\n");
        return 1;
    }

    HParseResult *result = h_parse(ntp_parser, buf, size);
    if (result) {
        print_ntp_packet(result);
        h_parse_result_free(result);
    } else {
        fprintf(stderr, "Parsing failed\n");
    }

    free(buf);
    fclose(f);
    return 0;
}

File: results/0.25/Network-Time-Protocol-Version-4/network-time-protocol-version-4-claude-3-5-sonnet-20241022.rs
use nom::{
    bits::complete::{tag, take},
    combinator::map,
    error::Error,
    sequence::tuple,
    IResult,
};
use std::{env, fs, path::Path};

#[derive(Debug)]
pub struct NTPPacket {
    li: u8,
    vn: u8,
    mode: u8,
    stratum: u8,
    poll: i8,
    precision: i8,
    root_delay: f32,
    root_dispersion: f32,
    reference_id: u32,
    reference_timestamp: u64,
    origin_timestamp: u64,
    receive_timestamp: u64,
    transmit_timestamp: u64,
    extensions: Vec<ExtensionField>,
    mac: Option<MAC>,
}

#[derive(Debug)]
pub struct ExtensionField {
    field_type: u16,
    length: u16,
    value: Vec<u8>,
}

#[derive(Debug)]
pub struct MAC {
    key_id: u32,
    digest: Vec<u8>,
}

fn parse_first_byte(input: (&[u8], usize)) -> IResult<(&[u8], usize), (u8, u8, u8)> {
    tuple((
        take(2usize),  // LI
        take(3usize),  // VN
        take(3usize),  // Mode
    ))(input)
}

fn parse_fixed_point_16_16(input: &[u8]) -> f32 {
    let integer = ((input[0] as u32) << 8 | input[1] as u32) as i32;
    let fraction = ((input[2] as u32) << 8 | input[3] as u32) as f32 / 65536.0;
    integer as f32 + fraction
}

fn parse_timestamp(input: &[u8]) -> u64 {
    let seconds = ((input[0] as u64) << 24)
        | ((input[1] as u64) << 16)
        | ((input[2] as u64) << 8)
        | (input[3] as u64);
    let fraction = ((input[4] as u64) << 24)
        | ((input[5] as u64) << 16)
        | ((input[6] as u64) << 8)
        | (input[7] as u64);
    (seconds << 32) | fraction
}

fn parse_extension_field(input: &[u8]) -> Option<(ExtensionField, &[u8])> {
    if input.len() < 4 {
        return None;
    }
    
    let field_type = ((input[0] as u16) << 8) | (input[1] as u16);
    let length = ((input[2] as u16) << 8) | (input[3] as u16);
    
    if input.len() < length as usize {
        return None;
    }

    let value = input[4..length as usize].to_vec();
    let padding = (4 - (length % 4)) % 4;
    let next_input = &input[(length + padding) as usize..];

    Some((ExtensionField {
        field_type,
        length,
        value,
    }, next_input))
}

fn parse_mac(input: &[u8]) -> Option<MAC> {
    if input.len() < 24 {
        return None;
    }

    let key_id = ((input[0] as u32) << 24)
        | ((input[1] as u32) << 16)
        | ((input[2] as u32) << 8)
        | (input[3] as u32);
    
    let digest = input[4..24].to_vec();

    Some(MAC { key_id, digest })
}

fn parse_ntp_packet(input: &[u8]) -> Option<NTPPacket> {
    if input.len() < 48 {
        return None;
    }

    let (_, (li, vn, mode)) = parse_first_byte((&input[0..1], 0)).ok()?;
    
    let mut packet = NTPPacket {
        li,
        vn,
        mode,
        stratum: input[1],
        poll: input[2] as i8,
        precision: input[3] as i8,
        root_delay: parse_fixed_point_16_16(&input[4..8]),
        root_dispersion: parse_fixed_point_16_16(&input[8..12]),
        reference_id: ((input[12] as u32) << 24)
            | ((input[13] as u32) << 16)
            | ((input[14] as u32) << 8)
            | (input[15] as u32),
        reference_timestamp: parse_timestamp(&input[16..24]),
        origin_timestamp: parse_timestamp(&input[24..32]),
        receive_timestamp: parse_timestamp(&input[32..40]),
        transmit_timestamp: parse_timestamp(&input[40..48]),
        extensions: Vec::new(),
        mac: None,
    };

    let mut remaining = &input[48..];
    while remaining.len() >= 4 {
        if let Some((extension, next_input)) = parse_extension_field(remaining) {
            packet.extensions.push(extension);
            remaining = next_input;
        } else {
            break;
        }
    }

    if !remaining.is_empty() {
        packet.mac = parse_mac(remaining);
    }

    Some(packet)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <ntp_packet_file>", args[0]);
        std::process::exit(1);
    }

    let path = Path::new(&args[1]);
    let data = match fs::read(path) {
        Ok(data) => data,
        Err(e) => {
            eprintln!("Error reading file: {}", e);
            std::process::exit(1);
        }
    };

    match parse_ntp_packet(&data) {
        Some(packet) => println!("{:#?}", packet),
        None => eprintln!("Failed to parse NTP packet"),
    }
}